File: leetcode/202-happy-number.js
// https://leetcode.com/problems/happy-number/ (top-interview-questions)

/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function(n) {
  
  // helper function...
  const sumOfDigitSquares = (num) => {
    let result = 0;
    
    while(num > 0) {
      result += Math.pow(num % 10, 2);
      num = Math.floor(num / 10);
    }

    return result;
  }

  const visitedNumbers = new Set();

  while (!visitedNumbers.has(n)) {
    visitedNumbers.add(n);
    n = sumOfDigitSquares(n);
    if (n === 1) {
      return true;
    }
  }

  return false;
};

File: leetcode/202-happy-number.py
# https://leetcode.com/problems/happy-number/ (top-interview-questions)

class Solution:
    def isHappy(self, n: int) -> bool:
        visited = set()
        def sumOfDigitSquares(n):
            result = 0
            while n:
                digit = n % 10
                digit *= digit
                result += digit
                n //= 10
            return result

        while n not in visited:
            visited.add(n)
            n = sumOfDigitSquares(n)
            if n == 1: return True
        return False
